third_party/terraform/data_sources/data_source_google_compute_instance_serial_port.go

```diff
@@ -0,0 +1,68 @@
+package google
+
+import (
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func dataSourceGoogleComputeInstanceSerialPort() *schema.Resource {
+	return &schema.Resource{
+		Read: computeInstanceSerialPortRead,
+		Schema: map[string]*schema.Schema{
+			"port": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"instance": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"zone": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"project": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"contents": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func computeInstanceSerialPortRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+	d.Set("project", project)
+	zone, err := getZone(d, config)
+	if err != nil {
+		return err
+	}
+	d.Set("zone", zone)
+
+	port := int64(d.Get("port").(int))
+	output, err := config.clientCompute.Instances.GetSerialPortOutput(project, zone, d.Get("instance").(string)).Port(port).Do()
+	if err != nil {
+		return err
+	}
+	contents := output.Contents
+	// When we reach the end of the serial port output stream the contents comes back empty
+	for output.Contents != "" {
+		output, err = config.clientCompute.Instances.GetSerialPortOutput(project, zone, d.Get("instance").(string)).Port(port).Start(output.Next).Do()
```

Comment:
Might be worth adding a debug log line here. It could be helpful understanding if this happens infinitely or takes longer than expected.